test/test-unicode-fold.c

```diff
@@ -0,0 +1,93 @@
+                    (int)_len,                                                                                         \
+                    _buf);                                                                                             \
+        ASSERT_CMPSIZE_T(_len, ==, expected_len);                                                                      \
+        ASSERT_CMPBYTES((uint8_t *)_buf, _len, (uint8_t *)expected, expected_len);                                     \
+        ASSERT_CMPUINT8((uint8_t)(_buf[_len]), ==, 0);                                                                 \
+        bson_free(_buf);                                                                                               \
+    } while (0)
+
+#define TEST_UNICODE_FOLD_ALL_CASES(input, case_folded, dia_folded, both_folded)                                       \
+    do {                                                                                                               \
+        size_t _input_len = strlen(input);                                                                             \
+        size_t _cf_len = strlen(case_folded);                                                                          \
+        size_t _df_len = strlen(dia_folded);                                                                           \
+        size_t _both_len = strlen(both_folded);                                                                        \
+        TEST_UNICODE_FOLD(case_folded, _cf_len, input, _input_len, kUnicodeFoldToLower);                               \
+        TEST_UNICODE_FOLD(dia_folded, _df_len, input, _input_len, kUnicodeFoldRemoveDiacritics);                       \
+        TEST_UNICODE_FOLD(both_folded,                                                                                 \
+                          _both_len,                                                                                   \
+                          input,                                                                                       \
+                          _input_len,                                                                                  \
+                          kUnicodeFoldToLower | kUnicodeFoldRemoveDiacritics);                                         \
+    } while (0)
+
+static void test_unicode_fold(_mongocrypt_tester_t *tester) {
+    mongocrypt_status_t *status = mongocrypt_status_new();
+    // Test all ascii chars.
+    char *buf1 = bson_malloc0(2);
+    char *buf2 = bson_malloc0(2);
+    for (unsigned char ch = 0; ch <= 0x7f; ch++) {
+        buf1[0] = ch;
+        // Diacritics
+        if (ch >= 'A' && ch <= 'Z') {
+            buf2[0] = ch + 0x20;
+            TEST_UNICODE_FOLD_ALL_CASES(buf1, buf2, buf1, buf2);
+        } else if (ch == '^' || ch == '`') {
+            TEST_UNICODE_FOLD_ALL_CASES(buf1, buf1, "", "");
+        }
```

Comment:
Add an `else` block to test all other single-byte cases. We just need to make sure they don't expand to >2 bytes, since that's assumed by the memory allocation in `unicode_fold`.